### PR TITLE
Fixes and additions

### PR DESCRIPTION
**Pull request by apaszke**
* Fixes segfault when `torch.Size` is constructed with invalid arguments (#555).
* Improve optimizer serialization and add `load_state_dict` (#549).
* Make `Variable`s non-comparable (#538).
* Add upsampling modules and functions to nn (#550)

### CODE REVIEW

File: torch/csrc/Size.cpp [hunk 1]
```diff
@@ -24,18 +24,17 @@ PyObject * THPSize_New(int dim, long *sizes)
 
 static PyObject * THPSize_pynew(PyTypeObject *type, PyObject *args, PyObject *kwargs)
```
Comment (colesbury) [inline torch/csrc/Size.cpp]: Add a test for this please

Comment (apaszke) [inline torch/csrc/Size.cpp]: Done

File: torch/autograd/variable.py [hunk 2]
```diff
@@ -741,6 +741,27 @@ def __len__(self):
     def __iter__(self):
         return iter(map(lambda i: self[i], range(self.size(0))))
 
+    def __eq__(self, other):
+        return NotImplemented
```
Comment (colesbury) [inline torch/autograd/variable.py]: At some point in the not too distant future we should implement these even if they return outputs that are marked as "not differentiable"

File: torch/autograd/variable.py [hunk 3]
```diff
@@ -741,6 +741,27 @@ def __len__(self):
...
+    def __le__(self, other):
+        return NotImplemented
+
+    def __gt__(self, other):
+        return NotImplemented
+
+    def __ge__(self, other):
+        return NotImplemented
+
+    def __hash__(self):
```
Comment (colesbury) [inline torch/autograd/variable.py]: What's wrong with the built-in hash function?

**Review by colesbury — APPROVED** [review]
lgtm

File: torch/autograd/variable.py (hunk 3 above)
Comment (apaszke) [inline torch/autograd/variable.py]: Nothing, but adding `__eq__` and not `__hash__` makes the type unhashable.

File: torch/autograd/variable.py (hunk 2 above)
Comment (apaszke) [inline torch/autograd/variable.py]: Yes, definitely. I just wanted to fix the issue for now.

File: torch/optim/sgd.py [hunk 4]
```diff
@@ -46,7 +46,7 @@ def step(self, closure=None):
                 if weight_decay != 0:
                     d_p.add_(weight_decay, p.data)
                 if momentum != 0:
-                    param_state = self.state[id(p)]
+                    param_state = self.state[p]
```
Comment (szagoruyko) [inline torch/optim/sgd.py]: breaks a lot of my code eg here https://github.com/szagoruyko/attention-transfer/blob/master/cifar.py#L314 but ok, my fault for not resetting the optimizer..

**Pull request merged**